content/blog/2025-05-06-typed-based-alias-analysis.md

```diff
@@ -0,0 +1,114 @@
+First, the authors use type information to compare two fields. If the types of two fields are not the same, and they also have no subtypes in common, then by the semantics of most languages, they cannot alias the same memory (at least, not without triggering undefined behavior).
+Second, the authors extend the analysis with field names. Even for two objects of the same type, accessing two different fields on two different objects cannot be aliases of the same memory.
+Third, the authors extend the analysis again by also performing a (flow-insensitive) analysis of assignments, to check if a common subtype is ever assigned to variables of two different parent types. If not, then any variables of those two parent types cannot be aliases either, despite sharing a subtype.
+The authors show that their later analyses are within 2.5% of a theoretically perfect analysis when applied to Redundant Load Elimination. One of the main contributions they emphasize is how they also performed a dynamic and limit analysis, in addition to the static analysis.
+
+## Annotations, Rich Types, and Alias Analysis
+
+
+One of the aspects of alias analysis is that it underestimates the true count of aliasing in a program. In TBAA it might be the case that we write a function that may alias but that in the total usages of the function it never actually aliases. In this scenario, we lose the opportunity to optimize some code, but we accept this tradeoff in order to keep the analysis fast and correct. In particular, TBAA is really powerful as it only relies on language constructs to perform the analysis so it doesn’t require special modifications to the source code to improve the results of the alias analysis. But this same approach opens more doors for optimizations if the source language has a richer type system or if there is a way for programmers to provide hints to the compiler. An example is the restrict keyword in C, where the programmers inform the compiler there will be no aliasing between parameters: 
+```C
+//source code
+bool myfn(int* restrict a, int* restrict b) {
+	*a = 1;
+	*b = 2;
+	return *a == *b;
+}
+
+// optimized version
+bool myfn(int* restrict a, int* restrict b) {
+	*a = 1;
+	*b = 2;
+	return false;
+}
+```
+
+In languages with rich type systems the type of the parameter might provide enough information to the compiler so that annotations are not necessary. For example, if we define a function in C++ that receives [unique pointers](https://en.cppreference.com/w/cpp/memory/unique_ptr), we can achieve the same optimization without the need for the use of a restrict keyword: 
+
+```C++
+//source code
+bool myfn(std::unique_ptr<int> a, std::unique_ptr<int> b) {
+	*a = 1;
+	*b = 2;
+	return *a == *b;
+}
+
+// optimized version
+bool myfn(std::unique_ptr<int> a, std::unique_ptr<int> b) {
+	return false;
+}
+```
```

Comment:
Just to confirm: is this a hypothetical, or did you actually try this in a real C++ compiler? Hypothetical is totally fine, but if you happen to have tried it out, that would be nice to mention explicitly (along with the compiler name & version you used).